jaxrs/WFLY-16018_EAP7-627.adoc

```diff
@@ -0,0 +1,107 @@
+:author:            Wei Nan Li
+:email:             weli@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.redhat.com/browse
+
+== Overview
+
+RESTEasy should be able to provide trace information about request in response headers. These trace information should be in headers if:
+
+it is required by global settings for all application (for example by EAP7-634)
+it is required by settings for one application (for example by EAP7-634)
+it is required in request header (for example: curl -v --header "RESTEasy-trace: true" http://localhost:8080/jaxrs-wf/some)
+
+Trace information should contains all necessary information and times about processing of current request, like used interceptors, used exception mappers, used end-points, used providers, etc. Example:
+
+[source,bash]
+----
+$ curl -v --header "RESTEasy-trace: true" http://localhost:8080/jaxrs-wf/some
+*   Trying ::1...
+* connect to ::1 port 8080 failed: Connection refused
+*   Trying 127.0.0.1...
+* Connected to localhost (127.0.0.1) port 8080 (#0)
+> GET /jaxrs-wf/some HTTP/1.1
+> Host: localhost:8080
+> User-Agent: curl/7.43.0
+> Accept: */*
+> RESTEasy-trace: true
+>
+< HTTP/1.1 200 OK
+< Connection: keep-alive
+< X-Powered-By: Undertow/1
+< Server: JBoss-EAP/7
+< Content-Type: application/octet-stream
+< Content-Length: 22
+< Date: Thu, 13 Apr 2017 08:02:44 GMT
+< RESTEasy-trace: 09:54:06,535 TRACE [org.jboss.resteasy.resteasy_jaxrs.whatever]  request was received, path: http://localhost:8080/jaxrs-wf/some
+< RESTEasy-trace: 09:54:06,735 TRACE [org.jboss.resteasy.resteasy_jaxrs.whatever]  MyCustomReaderInterceptor is used
+< RESTEasy-trace: 09:54:06,785 TRACE [org.jboss.resteasy.resteasy_jaxrs.whatever]  JAXB provider is used
+< RESTEasy-trace: 09:54:06,835 TRACE [org.jboss.resteasy.resteasy_jaxrs.whatever]  request is processed by MyCustomEndPointResource
+< RESTEasy-trace: 09:54:06,875 TRACE [org.jboss.resteasy.resteasy_jaxrs.whatever]  Jackson 2 provider is used
+< RESTEasy-trace: 09:54:06,935 TRACE [org.jboss.resteasy.resteasy_jaxrs.whatever]  MyCustomWriterInterceptor is used
+* Connection #0 to host localhost left intact
+{"x" : 5, "y" : "foo"}
+----
+
+== Issue Metadata
+
+=== Issue:
+
+* {issue-base-url}/WFLY-16018[WFLY-16018]
+
+=== Related Issues:
+
+* {issue-base-url}/EAP7-627[EAP7-627]
+* {issue-base-url}/EAP7-634[EAP7-634]
+* {issue-base-url}/EAP7-627[EAP7-606]
+
+=== Dev Contacts:
+
+* mailto:{email}[{author}]
+* mailto:jperkins@redhat.com[James Perkins]
+
+=== QE Contacts:
+
+TBD
+
+=== Testing By
+
+* [ ] Engineering
+* [ ] QE
+
+=== Affected Projects or Components:
+
+* WildFly
+* RESTEasy
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+* [x] Managed domain
+* [x] OpenShift s2i
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
```

Comment:
I added some requirements.

Reply:
@bstansberry Here is my reply:

> Can you fill in more details? AIUI there's meant to be some sort of management API support to allow this to be configured. What are the requirements for how that is done? Are there other ways to configure this beyond the management API?

Currently this module doesn't support management API configuration. Now the configuration is like this:

- If `resteasy-tracing` module is added into a project as dependency, then it will be loaded into runtime. By default it's turned off, and users can configure to use it by setting in `web.xml`. The other way to enable it is to use the `ON_DEMAND` mode(The usage is added into RESTEasy doc as QA required: https://github.com/resteasy/resteasy/pull/3235). Which the client side can use headers to control the tracing info.

> Per your note on the next line, this is a security sensitive configuration setting, so how will access to that be controlled via the WFLY management RBAC system?

For the security sensitive issue, it means the tracing info may contain RESTEasy server side internal messages (for example, the message passed into filters, the time information that consumed by each layer of internal RESTEasy classes, etc). Because it's a `tracing` feature, so it may expose a lot of internal tracing information when used.

> What is the required set of information that is exposed? That is, a formal list describing all that's meant to be covered under 'like used interceptors, used exception mappers, used end-points, used providers, etc.'

By default the feature is turned off, and if users want to use it, there are several levels of tracing information can be configured(https://docs.jboss.org/resteasy/docs/6.2.0.Beta1/userguide/html_single/index.html#d4e4526):

- SUMMARY
- TRACE
- INFO

Currently these three levels are hard coded into RESTEasy internal classes, maybe in the future we could provide more fine grained level information levels. Regardless which level, the tracing info is quite internal information.

The full list of the events traced by the module is listed here(like used interceptors, used exception mappers, used end-points, used providers, etc):

- https://docs.jboss.org/resteasy/docs/6.2.0.Beta1/userguide/html_single/index.html#d4e4591


> What's the message format? Is it configurable? What compatibility guarantees do we provide with respect to the format and the message details?

Currently it supports PLAIN-TEXT and JSON format. The example text of the tracing info is shown in the project document: https://docs.jboss.org/resteasy/docs/6.2.0.Beta1/userguide/html_single/index.html#d4e4566

@bstansberry wdyt about the above info? If it's okay guess I can add the above text into Hard Requirement.